core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java

```diff
@@ -137,6 +139,78 @@ public RewriteStrategy options(Map<String, String> options) {
     ).collect(Collectors.toList());
   }
 
+  protected long targetFileSize() {
+    return this.targetFileSize;
+  }
+
+  /**
+   * Determine how many output files to create when rewriting. We use this to determine the split-size
+   * we want to use when actually writing files to avoid the following situation.
+   * <p>
+   * If we are writing 10.1 G of data with a target file size of 1G we would end up with
+   * 11 files, one of which would only have 0.1g. This would most likely be less preferable to
+   * 10 files each of which was 1.01g. So here we decide whether to round up or round down
+   * based on what the estimated average file size will be if we ignore the remainder (0.1g). If
+   * the new file size is less than 10% greater than the target file size then we will round down
+   * when determining the number of output files.
+   * @param totalSizeInBytes total data size for a file group
+   * @return the number of files this strategy should create
+   */
+  protected long numOutputFiles(long totalSizeInBytes) {
+    if (totalSizeInBytes < targetFileSize) {
+      return 1;
+    }
+
+    long fileCountWithRemainder = LongMath.divide(totalSizeInBytes, targetFileSize, RoundingMode.CEILING);
+    if (LongMath.mod(totalSizeInBytes, targetFileSize) > minFileSize) {
+      // Our Remainder file is of valid size for this compaction so keep it
+      return fileCountWithRemainder;
+    }
+
+    long fileCountWithoutRemainder = LongMath.divide(totalSizeInBytes, targetFileSize, RoundingMode.FLOOR);
+    long avgFileSizeWithoutRemainder = totalSizeInBytes / fileCountWithoutRemainder;
+    if (avgFileSizeWithoutRemainder < 1.1 * targetFileSize) {
+      // Round down and distribute remainder amongst other files
+      return fileCountWithoutRemainder;
+    } else {
+      // Keep the remainder file
+      return fileCountWithRemainder;
+    }
+  }
+
+  /**
+   * Returns the smaller of our max write file threshold, and our estimated split size based on
+   * the number of output files we want to generate.
+   */
+  protected long splitSize(long totalSizeInBytes) {
+    long estimatedSplitSize = totalSizeInBytes / numOutputFiles(totalSizeInBytes);
+    return Math.min(estimatedSplitSize, writeMaxFileSize());
+  }
+
+  protected static long inputFileSize(List<FileScanTask> fileToRewrite) {
```

Comment:
This seems the only one static method. Is that on purpose?

Reply:
No reason, i'll remove it